public/app/plugins/datasource/prometheus/language_provider.ts

```diff
@@ -111,11 +115,54 @@ export default class PromQlLanguageProvider extends LanguageProvider {
     return defaultValue;
   };
 
+  requestViaWorker(url: string, hasObjectResponse?: false): Promise<string[]>;
+  requestViaWorker(url: string, hasObjectResponse?: true): Promise<PromMetricsMetadata>;
+  requestViaWorker(url: string, hasObjectResponse?: boolean): Promise<string[] | PromMetricsMetadata>;
+  requestViaWorker(url: string, hasObjectResponse = false): Promise<string[] | PromMetricsMetadata> {
+    return new Promise(resolve => {
+      let nodes: string[] | PromMetricsMetadata = hasObjectResponse ? {} : [];
+      const streamWorker = new (StreamJSONResponse as any)() as StreamJSONResponseWorker;
+
+      streamWorker.onmessage = e => {
+        if (e.data === JSON_STREAM_DONE) {
+          streamWorker.terminate();
+          resolve(nodes);
+        }
+
+        if (Array.isArray(nodes)) {
+          nodes = nodes.concat(e.data);
+        } else {
+          Object.assign(nodes, e.data);
+        }
+      };
+
+      streamWorker.onerror = error => {
+        resolve(nodes);
+      };
+
+      streamWorker.postMessage({
+        chunkSize: Math.ceil(this.lookupMetricsThreshold / 20),
+        headers: this.datasource.basicAuth ? { Authorization: this.datasource.basicAuth } : { 'X-Grafana-Org-Id': 1 },
```

Comment:
Yeah that's obviously exactly what I tried to do here. :)
Since how oboe does requests is precisely the point why I chose to use it, one can't just "hand off" the task of executing the request somehow.

Reply:
Well I mean I am not sure for example this part https://github.com/grafana/grafana/blob/40d195e4a71e9ffefb50ad4ed2439df285650790/public/app/core/services/backend_srv.ts#L397 seems like we are considering some headers we are not dealing with in the worker. And even though we may not use all of it with prometheus, may create some errors in the future. So I would rather use some single handleHeaders fucntion that would be used both in the worker and backend_srv to make sure both are handling the headers in the same way.

Reply:
We have to guarantee that the request can get through to the datasource, possibly through out datasource proxy. For that it needs to set the Org ID out of the user context, otherwise the request won't get through, or this will fail in other ways in a multi-org installation. I realize that it would be quite neat if the datasource could arrange for its own transport, but there is too much at stake if let each datasource do that independently.
I feel like the workerization should already prevent the sluggishness, and I'd be happy if the user has to wait a bit to see all results. What else do you get from streaming?

Reply:
@davkal I'm actually just now wrapping up on isolating and exposing the header transformation done by the backend service, should be ready to review in a few minutes. :)